### ISSUE TITLE
server demo recording causes server crash on map change

### ISSUE THREAD

**Issue by mittermichal** [etlegacy/etlegacy#2226]
how to reproduce:
- lower MAX_PACKETLEN in net_chan.c:
`#define MAX_PACKETLEN           120`
- run etlded
- `sv_autodemo 1`
- wait until demo recording starts
- `map radar`
- `map oasis`

stack:
```
SV_Netchan_Transmit sv_net_chan.c:268 // client is the ETLTV,both client->netchan_start_queue and client->netchan_end_queue are NULL
SV_SendMessageToClient sv_snapshot.c:848
SV_SendClientSnapshot sv_snapshot.c:952
SV_FinalCommand sv_init.c:1238
SV_SpawnServer sv_init.c:672
SV_Map_f sv_ccmds.c:146 // map oasis
Cmd_ExecuteString cmd.c:1016
Cbuf_Execute cmd.c:279
Com_Frame common.c:3452
Sys_GameLoop sys_main.c:1053
main sys_main.c:1181
```

**Comment by mittermichal** [comment]
if I leave MAX_PACKETLEN unchanged the crash is rather sporadic, it could be possible that lowering it might cause the crash in a different way. But I don't know how that queue stuff work

**Comment by mittermichal** [comment]
@ryzyk-krzysiek has suggestion to have `SV_DemoAutoDemoRecord` only in  `SV_Frame` and not in `SV_SpawnServer`

**Comment by mittermichal** [comment]
> @ryzyk-krzysiek has suggestion to have `SV_DemoAutoDemoRecord` only in `SV_Frame` and not in `SV_SpawnServer`

I have tried to only call `SV_DemoAutoDemoRecord` in `SV_Frame` but didn't help, maybe it needs to be called not in 1st frame or maybe Its issue with not properly removing the ETLTV client on map change

**Comment by Helldembez** [comment]
I started looking at this again and I managed to reproduce it two times without touching anything like @mittermichal did. I feel like its a timing issue when things get cleared before or after certain things. In both instances netchan_end_queue en netchan_start_queue were NULL. The first time it happened when I changed map after stopwatch round 2 intermission end passing a vote for a map. The second time was after stopwatch round 2 intermission end and writing `/quit`. I had a breakpoint in the if statement in SV_Netchan_Transmit the whole time and it only entered there before crashing.

Here are the stacktraces:

In both instances I could not get it to crash directly, but after loading maps multiple times plus actually completing a obj map in SW and then readying up in intermission directly before calling the vote map crash.

```
SV_Netchan_Transmit(client_s *, msg_t *) sv_net_chan.c:268
SV_SendMessageToClient(msg_t *, client_s *) sv_snapshot.c:848
SV_SendClientSnapshot(client_s *) sv_snapshot.c:952
SV_FinalCommand(const char *, qboolean) sv_init.c:1239
SV_SpawnServer(const char *) sv_init.c:672
SV_Map_f() sv_ccmds.c:146
Cmd_ExecuteString(const char *) cmd.c:1016
Cbuf_Execute() cmd.c:279
Com_Frame() common.c:3505
Sys_GameLoop() sys_main.c:1053
SDL_main(int, char **) sys_main.c:1181
main_getcmdline() SDL_windows_main.c:82
WinMain(HINSTANCE__ *, HINSTANCE__ *, char *, int) SDL_windows_main.c:114
invoke_main() 0x00000000005a684d
__scrt_common_main_seh() 0x00000000005a66d7
__scrt_common_main() 0x00000000005a656d
WinMainCRTStartup(void *) 0x00000000005a68c8
<unknown> 0x0000000076b600f9
<unknown> 0x0000000077b37bbe
<unknown> 0x0000000077b37b8e
```

```
SV_Netchan_Transmit(client_s *, msg_t *) sv_net_chan.c:268
SV_SendMessageToClient(msg_t *, client_s *) sv_snapshot.c:848
SV_SendClientSnapshot(client_s *) sv_snapshot.c:952
SV_FinalCommand(const char *, qboolean) sv_init.c:1239
SV_Shutdown(const char *) sv_init.c:1275
Com_Quit_f() common.c:465
Cmd_ExecuteString(const char *) cmd.c:1016
Cbuf_Execute() cmd.c:279
Com_Frame() common.c:3452
Sys_GameLoop() sys_main.c:1053
SDL_main(int, char **) sys_main.c:1181
main_getcmdline() SDL_windows_main.c:82
WinMain(HINSTANCE__ *, HINSTANCE__ *, char *, int) SDL_windows_main.c:114
invoke_main() 0x00000000005a684d
__scrt_common_main_seh() 0x00000000005a66d7
__scrt_common_main() 0x00000000005a656d
WinMainCRTStartup(void *) 0x00000000005a68c8
<unknown> 0x0000000076b600f9
<unknown> 0x0000000077b37bbe
<unknown> 0x0000000077b37b8e
```

**Comment by Helldembez** [comment]
So, the crash is solved by marking the ETLTV client entity as BOT. Moving the call to SV_Frame is better anyway IMO. The BOT flag however brings up a few problems. In SV_ClientEnterWorld the BOT flag gets removed if the client is democlient(which is set in SV_DemoConnectDummy) and has a `cg_etVersion` length greater than 0(which is set to ETLTV in SV_DemoConnectDummy as well). 

This is probably done for playback reasons, so a client allegedly doesn't join and play as if it were a bot. I'm not sure why it would/should since there are no network packages recorded of that client doing that.

Even if we fix this, the current code base tries to be mod agnostic. Which poses its own problems. @ryzyk-krzysiek tries to solve this in this branch: https://github.com/etlegacy/etlegacy/compare/master...ryzyk-krzysiek:etlegacy:server-demos 

The solution probably cant be fully mod agnostic as @ryzyk-krzysiek ran into some problems, which aren't clear yet.

So big question to all devs: Do we merge ryzyk his branch and take it from there or should we rethink the solution from the ground up? @ETLegacyTeam 

Also playback client / compatibility should be kept in mind, not just recording it successfully.

**Comment by Aciz** [comment]
Fixed in #2358
